cpp/src/parquet/encoding.cc

```diff
@@ -3033,11 +3062,226 @@ class RleBooleanDecoder : public DecoderImpl, virtual public BooleanDecoder {
+          }
+          const ByteArray src{view};
+          if (ARROW_PREDICT_TRUE(last_value_.empty())) {
+            last_value_ = view;
+            suffix_encoder_.Put(&src, 1);
+            prefix_length_encoder_.Put({static_cast<int32_t>(0)}, 1);
+            previous_len = src.len;
+          } else {
+            uint32_t j = 0;
+            while (j < std::min(previous_len, src.len)) {
+              if (last_value_[j] != view[j]) {
+                break;
+              }
+              j++;
+            }
+            previous_len = j;
+            prefix_length_encoder_.Put({static_cast<int32_t>(j)}, 1);
+
+            const uint8_t* suffix_ptr = src.ptr + j;
+            const uint32_t suffix_length = static_cast<uint32_t>(src.len - j);
+            last_value_ =
+                string_view{reinterpret_cast<const char*>(suffix_ptr), suffix_length};
+            const ByteArray suffix(suffix_length, suffix_ptr);
+            suffix_encoder_.Put(&suffix, 1);
+          }
+          return Status::OK();
+        },
+        []() { return Status::OK(); }));
+  }
+
+  ::arrow::BufferBuilder sink_;
+  DeltaBitPackEncoder<Int32Type> prefix_length_encoder_;
+  DeltaLengthByteArrayEncoder<ByteArrayType> suffix_encoder_;
+  string_view last_value_;
+};
+
+template <>
+inline void DeltaByteArrayEncoder<FLBAType>::Put(const FixedLenByteArray* src,
+                                                 int num_values) {
+  if (descr_->type_length() == 0) {
+    return;
+  }
+  for (int i = 0; i < num_values; ++i) {
+    // Write the result to the output stream
+    DCHECK(src[i].ptr != nullptr) << "Value ptr cannot be NULL";
+    PARQUET_THROW_NOT_OK(sink_.Append(src[i].ptr, descr_->type_length()));
+  }
+}
+
+template <>
+void DeltaByteArrayEncoder<FLBAType>::Put(const ::arrow::Array& values) {
+  if (!::arrow::is_fixed_size_binary(values.type_id())) {
+    throw ParquetException("Only FixedSizeBinaryArray and subclasses supported");
+  }
+  PutFixedLenByteArray(checked_cast<const ::arrow::FixedSizeBinaryArray&>(values));
+}
+
+template <typename DType>
+void DeltaByteArrayEncoder<DType>::Put(const ::arrow::Array& values) {
+  AssertBaseBinary(values);
+  if (::arrow::is_binary_like(values.type_id())) {
+    PutBinaryArray(checked_cast<const ::arrow::BinaryArray&>(values));
+  } else if (::arrow::is_large_binary_like(values.type_id())) {
+    PutBinaryArray(checked_cast<const ::arrow::LargeBinaryArray&>(values));
+  }
+}
+
+template <typename DType>
+void DeltaByteArrayEncoder<DType>::Put(const T* src, int num_values) {
+  if (num_values == 0) {
+    return;
+  }
+  ArrowPoolVector<int32_t> prefix_lengths(num_values);
+
+  if (ARROW_PREDICT_TRUE(last_value_.empty())) {
+    last_value_ = string_view{reinterpret_cast<const char*>(src[0].ptr), src[0].len};
+    suffix_encoder_.Put(&src[0], 1);
+    prefix_lengths[0] = 0;
+  }
+
+  for (int32_t i = 1; i < num_values; i++) {
```

Comment:
a bit confusing here, if `last_value_` is not empty, the first value would be ignored.

Reply:
Refactored.